### DISCUSSION TITLE
MQTT Topic von kWh in Wh umrechnen

### DISCUSSION THREAD

**Discussion by WolfImBusch** [solectrus/solectrus#3759] [Q&A] [answered]
Einen schönen guten Abend,

ich habe heute natürlich gleich an dem Beta-Test "Selbstdefinierte Verbraucher" teilgenommen ( #3749). Ist auch alles erstmal super - allerdings habe ich nach einer etwas längeren Suche bemerkt bzw. ist mir durch einen Blick in die infludb aufgefallen, dass meine Waschmaschine die verbrauchte Leistung in kWh per MQTT schickt und nicht als Wh, weshalb die Darstellung in Solectrus natürlich nicht klappen kann. Ich muss den Wert deshalb umrechnen resp. mit 1000 multiplizieren und - bin gescheitert ... :-)
Ich denke, ich muss am MQTT Collector ansetzen und habe natürlich zuerst die Dokumentation gelesen (https://docs.solectrus.de/referenz/mqtt-collector/topics.html). 

Im MQTT Explorer sieht das bei mir so aus
<img width="364" alt="Bildschirmfoto 2024-12-08 um 18 39 59" src="https://github.com/user-attachments/assets/b6739d5c-9ceb-49e5-a1de-99c05dd149f3">

In der Datei ".env" habe ich noch folgendes versucht

> MAPPING_11_TOPIC=homeassistant/sensor/washing_machine_stromverbrauch
> MAPPING_11_JSON_FORMULA="state*1000"
> MAPPING_11_MEASUREMENT=waschmaschine
> MAPPING_11_FIELD=power
> MAPPING_11_TYPE=float

Das hat leider nicht geklappt ... im MQTT Log sehe ich erst gar nicht mehr, dass das Tonic empfangen und in die influxdb geschrieben wird. 

Hat mir bitte jemand noch einen Tipp, was ich tun könnte?

Vielen Dank an Alle und viele Grüße

**Comment by ledermann** [comment]
Ich denke, es liegt an den fehlenden geschweiften Klammern. Das hier würde ich mal versuchen:

```
MAPPING_11_JSON_FORMULA="{state}*1000"
```

Wichtig ist auch, dass alle Variablen, also auch `MAPPING_11_JSON_FORMULA`, im `compose.yaml` im `environment`-Abschnitt des MQTT-Collectors mit aufgeführt werden.

**Reply by WolfImBusch** [reply]
Hallo Georg,
zuerst vielen Dank für Deinen Hinweis. Leider führte dieser auch nicht zum Ziel. Ich kann leider nicht sagen warum und ohne Logs wird Deine Glaskugel auch überfordert sein. Ich kenne mich mit den Topics/MQTT leider nicht besonders gut aus aber ich vermute, dass schlichtweg kein JSON übertragen wird und deshalb die Umrechnung scheitert. Ist aber nicht tragisch - ist nur die Waschmaschine (die ich ohnehin im Energie-Dashboard von Home Assistent sehe) - die großen/interessanten Verbraucher neben der Wallbox fange ich über Shelly's ein ... Nochmal Danke und viele Grüße!!!

**Reply by ledermann** [reply]
Ja, ohne Logs ist es immer schwierig. Der MQTT-Collector protokolliert genau, was er bekommt und was er damit macht, das kann sehr hilfreich sein.

Dein Hinweis, dass es möglicherweise kein JSON ist, sondern der Wert als einfacher String daherkommt, ist nützlich. Es ist tatsächlich so, dass der MQTT-Collector derzeitig Formeln nur für JSON erlaubt. Dies habe ich gerade verbessert.

Falls du das noch ausprobieren magst, auch wenn du das Thema schon abgehakt hast:

1) Wechsel des MQTT-Collectors auf das Docker-Tag `develop` 

2) Folgendes Mapping einstellen:

    ```env 
    MAPPING_11_TOPIC=homeassistant/sensor/washing_machine_stromverbrauch
    MAPPING_11_FORMULA="{value} * 1000"
    MAPPING_11_MEASUREMENT=waschmaschine
    MAPPING_11_FIELD=power
    MAPPING_11_TYPE=float
    ```

    Zu beachten sind zwei Dinge: Die Variable heißt `MAPPING_xx_FORMULA`, nicht `MAPPING_xx_JSON_FORMULA`. In der angegebenen Formel wird der Wert dann über `{value}` referenziert.

**Reply by WolfImBusch** [reply]
Hallo Georg!

Wow! Danke! Natürlich probiere ich das sehr gerne aus und ich habe die Waschmaschine gerade nochmal angeschmissen ... :-)

Um ehrlich zu sein hatte ich schon in das MQTT Log reingeschaut aber von diesem Topic war schlichtweg nix zu sein. Das ist jetzt anders ... :-)

```
solectrus-mqtt-collector-1  | # Message from 2024-12-15 19:12:40 +0100
solectrus-mqtt-collector-1  |   topic = homeassistant/sensor/washing_machine_stromverbrauch/state
solectrus-mqtt-collector-1  |   message = 0.1
solectrus-mqtt-collector-1  |   => waschmaschine:power = 100.0
```

Es sieht jetzt richtig richtig gut aus! Auch in der UI von Solectrus!!

Ich danke Dir und viele Grüße

**Reply by patricknitsch** [reply]
Hallo Georg,

besteht denn im oben genannten Beispiel auch die Möglichkeit, Variablen zu addieren? Als Beispiel den Wert von Mapping_11_Formula mit Mapping_12_Formula?

Hintergrund ist der, dass ich momentan Werte in Homeassistant summiere und dann über das entsprechende Topic bereitstelle. Ich möchte aber den Broker von HA auslagern und alle Werte der Shellys und Tasmota dahin schicken. Dann müsste die Berechnung außerhalb von HA passieren.

**Reply by ledermann** [reply]
Derzeitig ist das nicht möglich. Das Problem ist nämlich, dass die einzelnen Topics unabhängig voneinander hereinkommen, also nicht synchron sind. Denkbar ist auch, dass ein Topic im Sekundentakt kommt, das andere im Minutentakt.

Eine Addition (oder sonstige Rechenoperation) mit Werten aus verschiedenen Topics würde es erforderlich machen, dass sich der MQTT-Collector den letzten Wert eines jeden Topics merkt und bei Änderung eines der beteiligten Werte (oder periodisch) eine Neuberechnung vornimmt. Ein solches Feature würde mehr Unabhängigkeit bringen, weil man neben dem MQTT-Broker kein weiteres externes Tool nötig wäre.

Für den ioBroker habe ich mir genau eine solche Addition mal gebastelt:
https://github.com/orgs/solectrus/discussions/3782
Das wäre dann nicht mehr nötig.

Ich habe mal ein Issue angelegt:
https://github.com/solectrus/mqtt-collector/issues/305

Ob und wann ich das realisieren kann, lasse ich aber mal offen.

**Comment by Zeppelin-PV** [comment]
@WolfImBusch 
sorry, wenn ich mich hier kurz zu Wort melde
prinzipiell hatte ich das selbe Anliegen / Problem
wie bekomme ich ein Verbraucher von HomeAssistant zu Solectrus
in der Diskussion #3717 habe ich Dank Georg Schritt für Schritt den Weg und somit zur Lösung gefunden.
Mich wundert warum der Stromverbrauch, sprich die Energie (kWh), und nicht die Leistung (W) zu Solectrus übertragen werden soll.
Solectrus zeigt im Altuell die Leistung an und in der Zeit (Tag, Woche usw.) die verbrauchte Energie.
Mit welchem Adapter (Messsteckdose) wird Deine Waschmaschine gemessen ?
Fakt ist doch, dass alle Messadapter einen Leistungswert in W ausgeben.
... oder ist Dein Anliegen doch ein anderes ???

**Reply by WolfImBusch** [reply]
Hi @Zeppelin-PV ,
bei mir ist es anders ... 😄 Es handelt sich um Miele Geräte und dort greife ich die Energiewerte ("stromverbrauch") nicht über einen Mess-Adapter ab, sondern "direkt" über eine Miele Integration in Home Assistant, die den Verbrauch leider nur in kWH und nicht in Wh von den Geräten ausliest. Von dort - also Home Assistant - schicke ich dann die Werte über MQTT zu Solectrus. 

Aber tatsächlich habe ich Blödmann vielleicht ein "Einheiten-Thema". Ich bekomme nämlich von Miele die Energie in kWh und würde aber für Solectrus wahrscheinlich einen Leistungswert (kW oder Wh) benötigen. Ich trinke mal ein Bier und denke drüber nach ... 🤪
Viele Grüße

**Reply by ledermann** [reply, marked as answer]
SOLECTRUS benötigt auf jeden Fall die Leistung (in W), die Energie (in Wh oder kWh) wird dann über das Integral ausgerechnet.

**Reply by Zeppelin-PV** [reply]
habe mich gerade schnell über die Miele App und deren API schlau gemacht, da wird keine Leistung in Watt übertragen.
in der Miele 3rd Party API Device Capabilities List wird das zumindest nicht aufgeführt.
Mein Rat, stecke ein Shelly dazwischen und Deine Miele steht in Solectrus, ohne API, ohne HomeAssistant und ohne Probleme ;-) 

https://www.miele.com/developer/download.html
